renderers/css/actions.py

```diff
@@ -36,13 +36,16 @@ class OpiExecuteCommand(OpiAction):
 class OpiOpen(OpiAction):
     """Renderer for write PV actions."""
     ACTION_TYPE = 'OPEN_DISPLAY'
+    MACRO_ERROR = 'Macros must have key and value as strings ({}:{} violates this rule)'
 
     def render(self, actions_node, action_model):
         action_node = super(OpiOpen, self).render(actions_node, action_model)
         macros_node = et.SubElement(action_node, 'macros')
         parent_macros_node = et.SubElement(macros_node, 'include_parent_macros')
         parent_macros_node.text = 'true' if action_model._parent_macros else 'false'
         for key, value in action_model._macros.items():
+            if not utils.is_string(key) or not utils.is_string(value):
```

Comment:
You are checking the type is a string here, however won't this fail in certain circumstances.
'1' is a Python string and would pass the test, but it is still not a valid XML element name.
A class that defines __str__ may well produce valid XML, but would not be allowed here.